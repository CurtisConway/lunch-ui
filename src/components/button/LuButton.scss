@import '../../assets/sass/_variables';
@import '../../assets/sass/_mixins';

button {
  appearance:none;
  box-shadow:none;
  border:none;
  outline:none;
  background:none;
  padding:0;
  display:block;
}

.lu-button {
  background-color:$black;
  color:$white;
  text-decoration:none;
  font-size:$base;
  line-height:1em;
  vertical-align:middle;
  font-family: $defaultFont;
  height: $defaultInputHeight;
  min-height:$defaultInputHeight;
  border-radius:$defaultRadius;

  &.small {
    max-width:150px;
  }

  &.large {
    max-width:300px;
  }

  &.fluid {
    width:100%;
  }
}

a.lu-button {
  @include flexCenter();
  flex-basis:auto;
}