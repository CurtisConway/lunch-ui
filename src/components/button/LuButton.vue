<template>
  <component
    :is="getTagName"
    :to="to || $attrs.href"
    :class="defaultClassList"
    :style="defaultStyleList"
  >
    <slot></slot>
  </component>
</template>

<script>
import './LuButton.scss';
import colors from '../../assets/js/mixins/colors';

export default {
  name: 'LuButton',
  mixins: [colors],
  inheritAttrs: true,
  props: {
    to: {
      type: [String, Location],
      default: () => undefined,
    },
    size: {
      type: [String, Number],
      default: () => '',
    },
    color: {
      type: String,
      default: () => 'blue',
    },
    type: {
      type: String,
      default: () => 'solid',
    },
  },
  computed: {
    getTagName() {
      if (this.to != null || this.$attrs.href != null) {
        return 'router-link';
      }

      return 'button';
    },
    defaultClassList() {
      const classes = [];
      classes.push({
        'lu-button': true,
        large: this.size === 'large',
        small: this.size === 'small',
        fluid: !this.size,
      });
      if (this.type === 'outline') {
        classes.push(this.textColor);
        classes.push(this.borderColor);
      } else if (this.type === 'text') {
        classes.push(this.textColor);
      } else {
        classes.push(this.bgColor);
        classes.push(this.borderColor);
      }
      return classes;
    },
    defaultStyleList() {
      const styles = {};
      if (typeof this.size === 'number') {
        styles['max-width'] = `${this.size}px`;
        styles.width = `${this.size}px`;
      }
      return styles;
    },
  },
};
</script>
