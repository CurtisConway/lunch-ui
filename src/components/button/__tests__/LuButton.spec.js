import Vue from 'vue';
import Router from 'vue-router';

import LuButton from '../LuButton.vue';

import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import { compileToFunctions } from 'vue-template-compiler'

describe('LuButton.vue', () => {
  let mountFunction;
  let router = Router
  let localVue = typeof Vue;

  beforeEach(() => {
    router = new Router()
    localVue = createLocalVue()
    localVue.use(Router)

    mountFunction = (options = {}) => {
      return mount(LuButton, {
        localVue,
        router,
        ...options,
      })
    }
  })

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot()
  });

  it('should render component with a type prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        type: 'solid',
      },
    }).html()).toMatchSnapshot()

    expect(mountFunction({
      propsData: {
        type: 'text',
      },
    }).html()).toMatchSnapshot()

    expect(mountFunction({
      propsData: {
        type: 'outline',
      },
    }).html()).toMatchSnapshot()
  });

  it('should render component with a color prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        type: 'solid',
        color: 'green',
      },
    }).html()).toMatchSnapshot()

    expect(mountFunction({
      propsData: {
        color: 'green',
        type: 'text',
      },
    }).html()).toMatchSnapshot()

    expect(mountFunction({
      propsData: {
        color: 'green',
        type: 'outline',
      },
    }).html()).toMatchSnapshot()
  });
});
