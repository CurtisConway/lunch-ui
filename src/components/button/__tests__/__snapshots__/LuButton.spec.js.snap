// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuButton.vue should render component and match snapshot 1`] = `<button class="lu-button fluid bg-blue border-blue"></button>`;

exports[`LuButton.vue should render component with a color prop and match snapshot 1`] = `<button class="lu-button fluid bg-green border-green"></button>`;

exports[`LuButton.vue should render component with a color prop and match snapshot 2`] = `<button class="lu-button fluid text-green"></button>`;

exports[`LuButton.vue should render component with a color prop and match snapshot 3`] = `<button class="lu-button fluid text-green border-green"></button>`;

exports[`LuButton.vue should render component with a type prop and match snapshot 1`] = `<button class="lu-button fluid bg-blue border-blue"></button>`;

exports[`LuButton.vue should render component with a type prop and match snapshot 2`] = `<button class="lu-button fluid text-blue"></button>`;

exports[`LuButton.vue should render component with a type prop and match snapshot 3`] = `<button class="lu-button fluid text-blue border-blue"></button>`;
